src/ray/core_worker/pubsub/publisher.h

```diff
@@ -99,39 +109,75 @@ class Subscriber {
   /// Testing only. Return true if there's no metadata remained in the private attribute.
   bool AssertNoLeak() const;
 
+  /// Return true if the subscriber is disconnected (if the subscriber is dead).
+  /// The subscriber is considered to be dead if there was no long polling connection for
+  /// the timeout.
+  bool IsDisconnected() const;
+
+  /// Return true if there was no new long polling connection for a long time.
+  bool IsActiveConnectionTimedOut() const;
+
  private:
   /// Cached long polling reply callback.
+  /// It is cached whenever new long polling is coming from the subscriber.
+  /// It becomes a nullptr whenever the long polling request is replied.
   LongPollConnectCallback long_polling_reply_callback_ = nullptr;
   /// Queued messages to publish.
   std::list<ObjectID> mailbox_;
+  /// Callback to get the current time.
+  const std::function<double()> get_time_ms_;
+  /// The time in which the connection is considered as timed out.
+  uint64_t connection_timeout_ms_;
   /// The maximum number of objects to publish for each publish calls.
   const uint64_t publish_batch_size_;
+  /// The last time long polling was connected in milliseconds.
+  double last_connection_update_time_ms_;
 };
 
-/// Pubsub server.
+/// Protocol detail
+///
+/// - Subscriber always send a long polling connection as long as there are subscribed
+/// entries from the publisher.
+/// - Publisher caches the long polling request and reply whenever there are published
+/// messages.
+/// - Publishes messages are batched in order to avoid gRPC message limit.
+/// - Look at CheckDeadSubscribers for failure handling mechanism.
+///
 class Publisher {
  public:
   /// Pubsub coordinator constructor.
   ///
-  /// \param is_node_dead A callback that returns true if the given node id is dead.
-  explicit Publisher(std::function<bool(const NodeID &)> is_node_dead,
+  /// \param periodical_runner Periodic runner
```

Comment:
Nit: Could mention that this periodical runner is used to periodically check for dead subscribers, although that's pretty obvious when looking at the constructor code. Up to you!